### PR TITLE
Fix issue #5219: Feature: PR Review

### LINKED ISSUE #5219: Feature: PR Review
**What problem or use case are you trying to solve?**
OpenHands has a resolver module which resolves GitHub issues, writes comments and reads comments, creates branches, commits and PRs.

We want to add to it a PR review functionality.

**Describe the UX of the solution you'd like**

When a PR is labeled with "review-pr", a github job to review the PR will be executed.

**Implementation**

- Read openhands/resolver and all its py files.
- Read the prompts/resolve sub-directory and its .jinja files.
- Read the github workflow file at ./github/workflows/openhands-resolver.yml (note that this is from the repo root)
- Think carefully how the resolver job works, its triggers, the repo it works with, how it reads a PR, how it comments on a PR, how it finds github comments and the corresponding source code. You can take the time to write out everything you find out because it helps you think. It's fine if you're throughout and your output is long. Keep it technical, don't pontificate.

Once you're done and understand the existing code, lets implement the new functionality of the resolver:
- make another job in the openhands-resolver workflow, beside auto-fix: review-pr
- it should be triggered for the label "review-pr"
- it should check out the pr, read its description, read its diff, and comments if any
- it should write a clear and comprehensive review as a comment to the PR.

For this feature implementation, you also need to:
- write a prompt in the openhands/resolver/prompts sub-directory for this, similar with existing prompts there (you can create its own folder)
- implement the necessary code in a similar way with the existing resolve-issue.py or other files you read previously, if it makes sense for the review-pr functionality
- however, do not add code if the review-pr functionality has no use for it. For example, a review is a comment, it doesn't make sense to "guess success" for it.
- add unit tests for the python functions you add.

**IMPORTANT**
We work on python 3.12, and so you can use builtin list, dict, | None, not the older variants.
Do NOT delete comments from the existing code.
Keep the new code you write aligned with the way existing code is written, style, comments, and patterns, otherwise it triggers my ocd.

### PR DESCRIPTION
**Pull request by enyst**
**End-user friendly description of the problem this fixes or functionality that this introduces**

- [ ] Include this change in the Release Notes. If checked, you must provide an **end-user friendly** description for your change below

---
**Give a summary of what the PR does, explaining any non-trivial design decisions**

This pull request fixes #5219.

While the AI agent has described what appears to be a complete solution, the last message indicates that they are merely summarizing what was allegedly done, without providing evidence that the actual implementation was successfully merged or completed. The thread shows two attempted fixes that were unsuccessful, with one resulting in a branch creation that required manual intervention.

Key points indicating the issue is not resolved:
1. The action runs referenced ( 11984150812 and 11992993529) were triggered but there's no indication of successful completion
2. A branch was created (openhands-fix-issue-5219) suggesting the automated fix failed
3. The AI's final message is descriptive of what "has been done" but doesn't provide concrete evidence of successful implementation or merge
4. There's no confirmation from the repository maintainers or system that the changes were actually implemented and working

To consider this issue resolved, we would need to see:
1. Successful merge of the implemented changes
2. Confirmation that the new review-pr workflow is functional
3. Evidence that the actual code changes described in the summary were successfully implemented and deployed

It appears the issue is still open and requires either further automated attempts or manual intervention to complete the implementation.

Automatic fix generated by [OpenHands](https://github.com/All-Hands-AI/OpenHands/) 🙌

---
**Link of any specific issues this addresses**
Fix #5219 

---

To run this PR locally, use the following command:
```
docker run -it --rm   -p 3000:3000   -v /var/run/docker.sock:/var/run/docker.sock   --add-host host.docker.internal:host-gateway   -e SANDBOX_RUNTIME_CONTAINER_IMAGE=docker.all-hands.dev/all-hands-ai/runtime:5e3bd51-nikolaik   --name openhands-app-5e3bd51   docker.all-hands.dev/all-hands-ai/openhands:5e3bd51
```

### CODE REVIEW

**Comment by enyst** [comment]
@openhands-agent-exp We have a problem. You are on your own on this one.

> Run cd /tmp && python -m openhands.resolver.review_pr \
/opt/hostedtoolcache/Python/3.12.7/x64/bin/python: No module named openhands.resolver.review_pr
Error: Process completed with exit code 1.

Review the code of the resolver and think about this error and alternatives to fix it. You may find there is a similar fix related to the -exp suffix of the macro or -experimental suffix of the auto fix label.

If you can propose a fix, do it.
If you can't propose a fix, call the finish tool with a clear explanation.

When you're done: good night.

**Comment by openhands-agent** [comment]
New OpenHands update

File: .github/workflows/openhands-resolver.yml [hunk 1]
```diff
@@ -112,7 +112,12 @@ jobs:
 
       - name: Install OpenHands
         run: |
-          python -m pip install --upgrade -r requirements.txt
+          if [[ "${{ github.event.label.name }}" == "review-pr" ]]; then
```
Comment (enyst) [inline .github/workflows/openhands-resolver.yml:115]: @openhands-agent-exp  The only trigger of this step is 'review-pr' label, from what I see. Verify this. If so, you probably don't need the conditional here when installing, just install from github.

**Comment by enyst** [comment]
> 05:49:16 - openhands:INFO: resolve_issue.py:206 - AgentFinishAction(outputs={}, thought='The changes look good. Let me explain the fix:\n\n1. The error occurred because the workflow was trying to run `python -m openhands.resolver.review_pr` but the module wasn\'t installed in the Python environment.\n\n2. The workflow was using the PyPI package version of openhands (`openhands-ai`) which doesn\'t include the new `review_pr.py` module yet.\n\n3. The fix is to install openhands directly from GitHub when the `review-pr` label is used, similar to how it\'s done for experimental features. This ensures we get the latest version with the new module.\n\n4. The change modifies the "Install OpenHands" step to check for the `review-pr` label and install from GitHub in that case.\n\nThis fix should resolve the error by ensuring the `review_pr` module is available in the Python environment when needed.\n\nI\'ll finish the interaction since the fix is complete.', action='finish')

**Comment by openhands-agent** [comment]
New OpenHands update

**Comment by enyst** [comment]
Nah. When applying the label review-pr, our current code got:
> Run cd /tmp && python -m openhands.resolver.review_pr \
/opt/hostedtoolcache/Python/3.12.7/x64/bin/python: No module named openhands.resolver.review_pr
Error: Process completed with exit code 1.

Review the code of the resolver and think about this error and alternatives to fix it. You may find there is a similar fix related to the -exp suffix of the macro or -experimental suffix of the auto fix label.

If you can propose a fix, do it.
If you can't propose a fix, call the finish tool with a clear explanation.

**Comment by openhands-agent** [comment]
New OpenHands update

**Comment by enyst** [comment]
openhands-agent You still didn't get it. The latest comments were not fixed.

Review the whole code of this PR. When the workflow installs from github, it installs from the main branch. The main branch does not have our new addition of "review-pr" and its associated module, so the workflow fails.

Modify the workflow to install this PR branch. Remember it needs to both check it out and install.

**Comment by openhands-agent** [comment]
New OpenHands update

File: .github/workflows/openhands-resolver.yml [hunk 2]
```diff
@@ -45,6 +45,86 @@ permissions:
...
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          script: |
+            github.rest.issues.createComment({
+              issue_number: ${{ env.ISSUE_NUMBER }},
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: `[OpenHands](https://github.com/All-Hands-AI/OpenHands) started reviewing the PR! You can monitor the progress [here](https://github.com/${context.repo.owner}/${context.repo.repo}/actions/runs/${context.runId}).`
+            });
+
+      - name: Install OpenHands
```
Comment (enyst) [inline .github/workflows/openhands-resolver.yml:109]:
```suggestion
```

File: .github/workflows/openhands-resolver.yml [hunk 3]
```diff
@@ -45,6 +45,85 @@ permissions:
...
+            github.rest.issues.createComment({
+              issue_number: ${{ env.ISSUE_NUMBER }},
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: `[OpenHands](https://github.com/All-Hands-AI/OpenHands) started reviewing the PR! You can monitor the progress [here](https://github.com/${context.repo.owner}/${context.repo.repo}/actions/runs/${context.runId}).`
+            });
+
+        run: |
+          python -m pip install --upgrade pip
+          python -m pip install --upgrade -r requirements.txt
```
Comment (enyst) [inline .github/workflows/openhands-resolver.yml:111]:
```suggestion
```

File: .github/workflows/openhands-resolver.yml [hunk 4]
```diff
@@ -45,6 +45,84 @@ permissions:
...
+          script: |
+            github.rest.issues.createComment({
+              issue_number: ${{ env.ISSUE_NUMBER }},
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: `[OpenHands](https://github.com/All-Hands-AI/OpenHands) started reviewing the PR! You can monitor the progress [here](https://github.com/${context.repo.owner}/${context.repo.repo}/actions/runs/${context.runId}).`
+            });
+
+        run: |
+          python -m pip install --upgrade pip
```
Comment (enyst) [inline .github/workflows/openhands-resolver.yml:110]:
```suggestion
```

File: .github/workflows/openhands-resolver.yml [hunk 5]
```diff
@@ -45,6 +45,83 @@ permissions:
...
+          github-token: ${{secrets.GITHUB_TOKEN}}
+          script: |
+            github.rest.issues.createComment({
+              issue_number: ${{ env.ISSUE_NUMBER }},
+              owner: context.repo.owner,
+              repo: context.repo.repo,
+              body: `[OpenHands](https://github.com/All-Hands-AI/OpenHands) started reviewing the PR! You can monitor the progress [here](https://github.com/${context.repo.owner}/${context.repo.repo}/actions/runs/${context.runId}).`
+            });
+
+        run: |
```
Comment (enyst) [inline .github/workflows/openhands-resolver.yml:109]:
```suggestion
```

**Comment by openhands-agent** [comment]
New OpenHands update

**Comment by enyst** [comment]
We have fixed this PR and it is, basically, working. See above an example of PR Review made with it.

**IMPORTANT**
Now we need to remember the important specifications for this PR:
> Keep the new code you write aligned with the way existing code is written, style, comments, and patterns, otherwise it triggers my ocd.

For example, this PR has lines like this:
> assert llm_model is not None  # For type checker

"For type checker", really? Yeah it may be true but it's not really pretty. Look up how this kind of tweak is made in the original resolver code. Pick up the patterns of the original feature in the codebase. And come back to apply them here.

I mean it. Style matters.

**Comment by openhands-agent** [comment]
New OpenHands update

**Pull request closed without merging**